src/purple/mod.rs

```diff
@@ -66,3 +70,56 @@ where
     let closure = &*(user_data as *mut F);
     closure(df, cond);
 }
+
+pub fn register_cmd<F>(
+    cmd: &str,
+    args: &str,
+    help_text: &str,
+    callback: F,
+) -> purple_sys::PurpleCmdId
+where
+    F: Fn(&mut ChatConversation, &str, Vec<&str>) -> purple_sys::PurpleCmdRet + 'static,
+{
+    let user_data = Box::into_raw(Box::new(callback)) as *mut c_void;
+    let c_cmd = CString::new(cmd).unwrap();
+    let c_args = CString::new(args).unwrap();
+    let c_help = CString::new(help_text).unwrap();
+
+    unsafe {
+        purple_sys::purple_cmd_register(
+            c_cmd.as_ptr(),
+            c_args.as_ptr(),
+            purple_sys::PurpleCmdPriority::PURPLE_CMD_P_DEFAULT,
+            purple_sys::PurpleCmdFlag::PURPLE_CMD_FLAG_CHAT,
+            ptr::null(),
+            Some(trampoline_cmd::<F>),
+            c_help.as_ptr(),
+            user_data,
+        )
+    }
+}
+
+unsafe extern "C" fn trampoline_cmd<F>(
+    conversation_ptr: *mut purple_sys::PurpleConversation,
+    c_cmd: *const c_char,
+    c_args: *mut *mut c_char,
+    _c_error: *mut *mut c_char,
+    user_data: *mut c_void,
+) -> purple_sys::PurpleCmdRet
+where
+    F: Fn(&mut ChatConversation, &str, Vec<&str>) -> purple_sys::PurpleCmdRet,
+{
+    let closure = &*(user_data as *mut F);
+
+    let cmd = CStr::from_ptr(c_cmd).to_str().unwrap();
+    let args: Vec<String> = FromGlibPtrContainer::from_glib_none(c_args);
+    // TODO: This is a conversation. Not necessairly a ChatConersation.
+    let mut conversation =
+        ChatConversation::from_ptr(conversation_ptr as *mut purple_sys::PurpleConvChat).unwrap();
+
+    closure(
+        &mut conversation,
+        cmd,
+        args.iter().map(std::ops::Deref::deref).collect(),
```

Comment:
you can consume `args` with `into_iter()` if you don't plan to use `args` afterward.

Reply:
Hmm I can't seem to do it.

I tried a couple of things but I always end up getting something similar:

```
123 |         args.into_iter().map(|s| s.as_str()).collect(),
    |                                  -^^^^^^^^^
    |                                  |
    |                                  returns a value referencing data owned by the current function
    |                                  `s` is borrowed here
```

Reply:
Make sense!
Nvm the `into_iter` :)